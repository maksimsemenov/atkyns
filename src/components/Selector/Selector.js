import React, { PropTypes } from 'react'
import classNames from 'classnames'
import l from 'utils/local'
import './Selector.less'

const Selector = ({ options = [], active = '', onChange }) => {
  const activeOptions = options.filter(option => !option.disabled)
  active = activeOptions.includes(active) ? active : activeOptions[0]
  return (
    <ul className='selector'>
      {options.map((option, index) => {
        const optionClasses = classNames({
          'selector__option': true,
          'is-active': option.name === active,
          'is-disabled': options.disabled
        })
        return (
          <li
            key={index}
            className={optionClasses}
            onClick={() => onChange(option)}
          >
            {l(`%${option.value}`)}
          </li>
        )
      })}
    </ul>
  )
}

Selector.propTypes = {
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.string.isRequired,
    disabled: PropTypes.bool
  })).isRequired,
  active: PropTypes.string,
  onChange: PropTypes.func.isRequired
}

export default Selector
