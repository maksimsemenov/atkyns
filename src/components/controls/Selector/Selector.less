@import '~constants/styles.less';

.selector__control {
  display: inline-block;
  max-width: 100%;
}
.selector {
  box-sizing: border-box;
  padding: 0;
  margin: 0;
  list-style: none;
  overflow: hidden;
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  align-items: center;

  border-radius: @radius-control;
  border: 1px solid @color-gray;
  font-size: @font-size-form;
  font-weight: @font-weight-basic;

  transition: border-color @transition-basic;

  &:hover {
    border-color: @color-dark-gray;
    transition: border-color @transition-active;
  }
  &.has-value {
    border-color: @color-accent;
    transition: border-color @transition-active;

    &:hover {
      border-color: @color-accent-active;
    }
  }
  &.has-error {
    border-color: @color-red;
    box-shadow: 0 0 3px 0 fade(@color-red, 30%);
  }
}

.selector__option {
  padding: 10px 24px 12px;
  cursor: pointer;

  color: @font-color-basic;
  border-right: 1px solid @color-gray;
  transition: border-color @transition-basic, background-color @transition-basic, color @transition-basic;

  .selector:hover & {
    border-color: @color-dark-gray;
  }
  &:hover {
    background-color: fade(@color-gray, 40);
    transition: background-color @transition-active
  }
  .selector.has-value & {
    border-color: @color-accent;
    color: @color-accent;
    &.is-active {
      color: #fff;
    }
  }
  .selector.has-value:hover & {
    border-color: @color-accent-active;
    transition: border-color @transition-active
  }
  .selector.has-value:hover &:hover {
    background-color: fade(@color-accent, 7);
    transition: background-color @transition-active;
  }
  &.is-active {
    pointer-events: none;
    background-color: @color-accent;
    border-color: @color-accent;
    color: #fff;
    transition: border-color @transition-active, background-color @transition-active, color @transition-active;

    &:hover,
    .selector.has-value:hover &:hover {
      background-color: @color-accent;
    }
  }
  .selector.has-error & {
    border-color: @color-red;
    color: @color-red;
    &:hover {
      border-color: @color-red;
      background-color: fade(@color-red, 7);
    }
  }
  &:last-child {
    border-right: none;
  }
}
