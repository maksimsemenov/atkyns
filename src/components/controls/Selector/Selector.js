import React, { PropTypes } from 'react'
import classNames from 'classnames'
import './Selector.less'

const Selector = ({ options = [], value = '', onChange, error, note, className }) => {
  const selectorClasses = classNames({
    'selector': true,
    'has-value': value,
    'has-error': error,
    [className]: className
  })
  return (
    <div className='selector__container'>
      <ul className={selectorClasses}>
        {options.map((option, index) => {
          const optionClasses = classNames({
            'selector__option': true,
            'is-active': option.value === value,
            'is-disabled': options.disabled
          })
          return (
            <li
              key={index}
              className={optionClasses}
              onClick={() => onChange(option.value)}
            >
              {option.name}
            </li>
          )
        })}
      </ul>
      {error ? <div className='selector__error'>{error}</div> : null}
      {!error && note && <div className='selector__note'>{note}</div>}
    </div>
  )
}

Selector.propTypes = {
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    disabled: PropTypes.bool
  })).isRequired,
  value: PropTypes.string,
  error: PropTypes.string,
  note: PropTypes.string,
  className: PropTypes.string,
  onChange: PropTypes.func.isRequired
}

export default Selector
