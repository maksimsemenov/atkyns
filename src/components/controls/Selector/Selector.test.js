import React from 'react'
import renderer from 'react-test-renderer'
import Selector from './Selector'

describe('Selector component', () => {
  it('renders without craches', () => {
    const options = [
      { name: 'Option 1', value: 'option1' },
      { name: 'Option 2', value: 'option2' },
      { name: 'Option 3', value: 'option3' }
    ]
    const onChange = () => {}
    const params = { options, onChange}
    const tree = renderer.create(<Selector {...params} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
