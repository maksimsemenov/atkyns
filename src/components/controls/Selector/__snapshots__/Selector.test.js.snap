exports[`Selector component renders without craches 1`] = `
<div
  className="selector__container">
  <ul
    className="selector">
    <li
      className="selector__option"
      onClick={[Function onClick]}>
      Option 1
    </li>
    <li
      className="selector__option"
      onClick={[Function onClick]}>
      Option 2
    </li>
    <li
      className="selector__option"
      onClick={[Function onClick]}>
      Option 3
    </li>
  </ul>
</div>
`;
